### PR TITLE
Dixu/launcher

### PR DESCRIPTION
**Pull request by xudifsd**
Support distributed/regular job in launcher now, but not inference job.

Still not fully tested. So I'll still use `PythonLauncher` as you can see in `job_manager.py`. So that the behavior should be unchanged.

Most of changes happen in:
* `job_launcher.py`, as I move many JobDeployer functions into super class of `PythonLauncer` and `LauncherStub`, it's meaningless to have a standalone JobDeployer class.
* `dist_pod_template.py` and `pod_template.py`, still have `generate_pods` behavior unchanged, so `PythonLauncher` will still works. But have an independent `generate_params` method, so that `framework.py` can be use.

I tried carefully do not change the behavior of old launcher. So we can have old one while developing new one.

I'd like to have this PR merged, so fewer conflicts will happen. After the merge I think I will add more test cases to ensure the behavior is expected in both old/new launcher, and then I will try to support inference job in new launcher.

Interesting data from testing:

With old launcher:
```
2020-01-30 19:54:53,231 : INFO : regular job 72978bb1-ae82-4187-a03a-faf99a1c665e created
2020-01-30 19:55:32,197 : INFO : spent 0:00:38.966011 in waiting job become running
2020-01-30 19:55:34,682 : INFO : killed regular job 72978bb1-ae82-4187-a03a-faf99a1c665e
2020-01-30 19:55:34,734 : INFO : distributed job 92ccd4d0-a2b3-4a89-acbe-ef38436c5af4 created
2020-01-30 19:56:17,946 : INFO : spent 0:00:43.211741 in waiting job become running
2020-01-30 19:56:20,490 : INFO : killed distributed job 92ccd4d0-a2b3-4a89-acbe-ef38436c5af4
2020-01-30 19:56:20,540 : INFO : data job 1134bdac-bf80-4f7c-a38d-221b4e8a66a5 created
2020-01-30 19:57:02,734 : INFO : spent 0:00:42.193177 in waiting job become finished
2020-01-30 19:57:02,795 : INFO : killed data job 1134bdac-bf80-4f7c-a38d-221b4e8a66a5
2020-01-30 19:57:02,875 : INFO : regular job 0a6e55fe-2c30-49ab-92ec-55b6136e4a44 created
2020-01-30 19:57:22,458 : INFO : spent 0:00:19.582393 in waiting job become failed
2020-01-30 19:57:22,487 : INFO : killed regular job 0a6e55fe-2c30-49ab-92ec-55b6136e4a44
```

With new launcher:
```
2020-01-30 19:51:36,550 : INFO : regular job ac2d0066-ed81-4d4a-863e-654b31c6dc34 created
2020-01-30 19:51:47,906 : INFO : spent 0:00:11.355929 in waiting job become running
2020-01-30 19:51:52,501 : INFO : killed regular job ac2d0066-ed81-4d4a-863e-654b31c6dc34
2020-01-30 19:51:52,535 : INFO : distributed job 99aee299-8395-4a74-94dd-217751e924e2 created
2020-01-30 19:52:03,844 : INFO : spent 0:00:11.309125 in waiting job become running
2020-01-30 19:52:06,529 : INFO : killed distributed job 99aee299-8395-4a74-94dd-217751e924e2
2020-01-30 19:52:06,562 : INFO : data job b35244b2-fcad-4938-812c-5a2f281d50f0 created
2020-01-30 19:52:50,911 : INFO : spent 0:00:44.349710 in waiting job become finished
2020-01-30 19:52:50,973 : INFO : killed data job b35244b2-fcad-4938-812c-5a2f281d50f0
2020-01-30 19:52:51,008 : INFO : regular job 83722121-736c-4d3c-9620-522fd9d318ed created
2020-01-30 19:53:10,637 : INFO : spent 0:00:19.628813 in waiting job become failed
2020-01-30 19:53:10,658 : INFO : killed regular job 83722121-736c-4d3c-9620-522fd9d318ed
```

new launcher have some speed up in waiting job running, more than 20s, I tried several times, same result. Not sure why, I expect the new launcher will be slower because a new component is introduced, but it's nice thing to have.

### CODE REVIEW

File: src/ClusterBootstrap/deploy.py [hunk 1]
```diff
@@ -2461,7 +2461,7 @@ def get_all_services():
                 with open(yamlname) as f:
                     content = f.read()
                     f.close()
-                    if content.find("Deployment") >= 0 or content.find("DaemonSet") >= 0 or content.find("ReplicaSet") >= 0 or content.find("CronJob") >= 0:
+                    if content.find("Deployment")>=0 or content.find("DaemonSet")>=0 or content.find("ReplicaSet")>=0 or content.find("CronJob")>=0 or content.find("StatefulSet")>=0:
```
Comment (Anbang-Hu) [inline src/ClusterBootstrap/deploy.py]: Mind adding back the spaces?

File: src/ClusterBootstrap/deploy.py [hunk 2]
```diff
@@ -2461,7 +2461,11 @@ def get_all_services():
                 with open(yamlname) as f:
                     content = f.read()
                     f.close()
-                    if content.find("Deployment") >= 0 or content.find("DaemonSet") >= 0 or content.find("ReplicaSet") >= 0 or content.find("CronJob") >= 0:
+                    if content.find("Deployment") >= 0 or \
+                            content.find("DaemonSet") >= 0 or \
+                            content.find("ReplicaSet") >= 0 or \
+                            content.find("CronJob") >= 0 or \
+                            content.find("StatefulSet") >= 0:
                         # Only add service if it is a daemonset.
```
Comment (Anbang-Hu) [inline src/ClusterBootstrap/deploy.py]: This comment seems outdated: `# Only add service if it is a daemonset.`

File: src/ClusterManager/job_manager.py [hunk 3]
```diff
@@ -350,7 +341,7 @@ def UpdateJobStatus(redis_conn, launcher, job, notifier=None, dataHandlerOri=Non
             dataFields = {
                 "jobStatusDetail": base64.b64encode(json.dumps(detail).encode("utf-8")).decode("utf-8"),
                 "jobStatus": "running"
-            }
+                }
```
Comment (Anbang-Hu) [inline src/ClusterManager/job_manager.py]: Indentation for `}` looks weird.

File: src/ClusterManager/pod_template.py [hunk 4]
```diff
@@ -16,10 +16,9 @@
 from osUtils import mkdirsAsUser
 
 class PodTemplate():
-    def __init__(self, template, deployment_template=None, enable_custom_scheduler=False, secret_templates=None):
+    def __init__(self, template, deployment_template=None, secret_templates=None):
```
Comment (Anbang-Hu) [inline src/ClusterManager/pod_template.py]: Let's remove `enable_custom_scheduler` in `DistPodTemplate` as well.

File: src/ClusterManager/job_launcher.py [hunk 5]
```diff
@@ -308,44 +467,14 @@ def get_secrets(self, field_selector="", label_selector=""):
...
-
     @record
     def pod_exec(self, pod_name, exec_command, timeout=60):
         """work as the command (with timeout): kubectl exec 'pod_name' 'exec_command'"""
         try:
             logger.info("Exec on pod {}: {}".format(pod_name, exec_command))
             client = stream(
-                self.k8s_CoreAPI.connect_get_namespaced_pod_exec,
-                name=pod_name,
+                    self.k8s_CoreAPI.connect_get_namespaced_pod_exec,
```
Comment (Anbang-Hu) [inline src/ClusterManager/job_launcher.py]: The indentation of this line and the next line doesn't look correct.

File: src/ClusterManager/job_launcher.py [hunk 6]
```diff
@@ -26,19 +26,197 @@
...
+        if "finishedAt" not in pod_status_detail:
+            pod_status_detail["finishedAt"] = finished_at
+
+        pod_status_detail["message"] = "{} at {}. {}".format(
+            status, finished_at, msg)
+        new_job_status_detail.append(pod_status_detail)
+
+    return new_job_status_detail
+
+def get_job_status_detail(job):
```
Comment (Anbang-Hu) [inline src/ClusterManager/job_launcher.py]: This function has been defined above on line 126

File: src/ClusterManager/job_launcher.py [hunk 7]
```diff
@@ -26,19 +26,197 @@
...
+        return job_status_detail
+
+    if not isinstance(job_status_detail, list):
+        job_status_detail = base64.b64decode(
+            job_status_detail.encode("utf-8")).decode("utf-8")
+        job_status_detail = json.loads(job_status_detail)
+    return job_status_detail
+
+
+def job_status_detail_with_finished_time(job_status_detail, status, msg=""):
```
Comment (Anbang-Hu) [inline src/ClusterManager/job_launcher.py]: This function has been defined above on line 141

File: src/ClusterManager/endpoint_manager.py [hunk 8]
```diff
@@ -39,10 +38,45 @@ def is_ssh_server_ready(pod_name):
         return False
     return True
 
+def pod_exec(self, pod_name, exec_command, timeout=60):
```
Comment (xudifsd) [inline src/ClusterManager/endpoint_manager.py]: should remove `self` argument

**Pull request merged**